0006-ffmul-revised.md

```diff
@@ -0,0 +1,2249 @@
+
+### `ForeignFieldMul` gadget
+
+As the soundness proof shows, the `ForeignFieldMul` gate is not complete without a number of external checks. The job of an `ForeignFieldMul` _gadget_ (a provable method which wraps the gate) is to take care of these checks and provide an API without pitfalls.
+
+As usual, we want the gadget to add all necessary checks to its outputs $q$ and $r$ but not to its inputs $a$ and $b$ (because, if all gadgets fully constrained their inputs, then we would add these constraints multiple times if we used the same inputs in multiple gadgets, which can be very inefficient).
+
+In addition, there should be an advanced flag to skip range checks on $r$, for use cases where $r$ is later asserted to equal a value which is already known to be in the range (for example, the constant 1 when constraining a foreign field inverse).
+
+By default, we will do even more checks on $r$ than required in our soundness proof: We also add a bounds check for $r$ which shows that $r < 2^{2\ell}(f_2 + 1)$. This enables us to use $r$ as input to another ffmul gate or other foreign field operations without extra checks, which is what we expect from a safe API.
+
+From these considerations, we propose the following logic of the ffmul gadget in pseudo-code:
+
+```ts
+function multiply(
+  a: ForeignField,
+  b: ForeignField,
+  externalChecks: ExternalChecks,
+  foreignModulus: bigint,
+  skipRemainderCheck?: boolean
+): {
+  q: ForeignField;
+  r: Option<ForeignField>;
+  rCompact: [Field, Field];
+} {
+  // compute witnesses
+  let { q, r01, r2, p10, p11, c0, c1, qBound } = exists(FFMulWitness, () => {
+    /* ...*/
+  });
+
+  // add the ffmul gate
+  addFFmulGate({ a, b, q, r01, r2, p10, p11, c0, c1, qBound }, foreignModulus);
+
+  // add 3 range checks on limbs of q
+  addMultiRangeCheck(q);
+
+  // add single range check on q bound to an accumulator for external checks
+  // (because these need to be in batches of 3 for efficiency)
+  externalChecks.addRangeCheck(qBound);
+
+  // if we're skipping the r checks, only return q and r in compact form
+  if (skipRemainderCheck) {
+    return { q, r: None, rCompact: [r01, r2] };
+  }
+
+  // split up the compact r01 limb to get witnesses for r range check
+  let [r0, r1] = exists([Field, Field], () => splitCompactLimb(r01));
+  let r = ForeignField.fromLimbs([r0, r1, r2]);
+
+  // add range check on r (compact mode)
+  addCompactRangeCheck(r, r01);
+
+  // compute r bound using provable native field addition
+  let [_f0, _f1, f2] = toLimbs(foreignModulus);
+  let rBound = r2.add(2n ** limbSize - f2 - 1n);
+
+  // add range check on r bound to external checks
+  externalChecks.addRangeCheck(rBound);
+
+  // return q, r, rCompact
+  return { q, r: Some(r), rCompact: [r01, r2] };
+}
+```
+
+### Spec
+
+[WIP Kimchi PR with spec change](https://github.com/o1-labs/proof-systems/blob/ffmul/rangecheck/book/src/specs/kimchi.md#foreign-field-multiplication) 
```

Comment:
The idea (mentioned in the RFC template under "Ephemeral details") was to link to a PR with spec changes from within the RFC. But since that code spec doesn't include more info than this RFC, I'm fine to delete